File: backend/app/schemas/maintenance.py
from __future__ import annotations
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
import uuid
from app.models.enums import ServiceType, ServiceStatus

class MaintenanceRecordCreate(BaseModel):
    service_type: ServiceType
    description: Optional[str] = None
    cost: Optional[float] = Field(default=None, ge=0)
    mileage_at_service: Optional[float] = Field(default=None, ge=0)
    performed_by: Optional[str] = Field(default=None, max_length=100)
    performed_at: date
    notes: Optional[str] = None

class MaintenanceRecordUpdate(BaseModel):
    description: Optional[str] = None
    cost: Optional[float] = Field(default=None, ge=0)
    mileage_at_service: Optional[float] = Field(default=None, ge=0)
    performed_by: Optional[str] = Field(default=None, max_length=100)
    performed_at: Optional[date] = None
    notes: Optional[str] = None

class MaintenanceRecordOut(BaseModel):
    id: uuid.UUID
    truck_id: uuid.UUID
    service_type: ServiceType
    description: Optional[str]
    cost: Optional[float]
    mileage_at_service: Optional[float]
    performed_by: Optional[str]
    performed_at: date
    notes: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True

class FuelLogCreate(BaseModel):
    liters: float = Field(gt=0)
    cost_per_liter: float = Field(gt=0)
    total_cost: Optional[float] = Field(default=None, gt=0)
    mileage_at_fill: Optional[float] = Field(default=None, ge=0)
    fuel_station: Optional[str] = Field(default=None, max_length=200)
    filled_at: Optional[datetime] = None
    trip_id: Optional[uuid.UUID] = None

class FuelLogOut(BaseModel):
    id: uuid.UUID
    truck_id: uuid.UUID
    trip_id: Optional[uuid.UUID] = None
    liters: float
    cost_per_liter: float
    total_cost: float
    mileage_at_fill: Optional[float]
    fuel_station: Optional[str]
    filled_at: datetime
    created_at: datetime

    class Config:
        from_attributes = True

class ServiceIntervalOut(BaseModel):
    id: uuid.UUID
    truck_id: uuid.UUID
    service_type: ServiceType
    interval_km: Optional[float]
    interval_days: Optional[int]
    last_service_date: Optional[date]
    last_service_mileage: Optional[float]
    next_service_date: Optional[date]
    next_service_mileage: Optional[float]
    status: ServiceStatus

    class Config:
        from_attributes = True

File: frontend/src/lib/maintenance.ts
import { api } from '@/lib/api';

export interface FuelSummaryTruck {
  truck_id: string;
  truck_name: string;
  plate_number: string;
  gallons: number;
  cost: number;
  miles: number;
  mpg: number;
}

export interface FuelSummary {
  days: number;
  total_cost: number;
  total_gallons: number;
  total_miles: number;
  fleet_mpg: number;
  avg_price_per_gallon: number;
  trucks: FuelSummaryTruck[];
}

export interface BackendServiceInterval {
  id: string;
  truck_id: string;
  service_type: string;
  interval_km: number | null;
  interval_days: number | null;
  last_service_date: string | null;
  last_service_mileage: number | null;
  next_service_date: string | null;
  next_service_mileage: number | null;
  status: 'scheduled' | 'overdue' | 'completed';
}

export interface BackendMaintenanceRecord {
  id: string;
  truck_id: string;
  service_type: string;
  description: string | null;
  cost: number | null;
  mileage_at_service: number | null;
  performed_by: string | null;
  performed_at: string;
  notes: string | null;
  created_at: string;
}

export interface BackendFuelLog {
  id: string;
  truck_id: string;
  liters: number;          // gallons in demo
  cost_per_liter: number;  // $/gallon in demo
  total_cost: number;
  mileage_at_fill: number | null;
  fuel_station: string | null;
  filled_at: string;
  created_at: string;
}

export interface MaintenanceReminder {
  interval: BackendServiceInterval;
  truck: { id: string; name: string; plate_number: string } | null;
  due_soon: boolean;
}

export const maintenanceApi = {
  fuelSummary: (days = 30) =>
    api<FuelSummary>(`/api/maintenance/fuel-summary?days=${days}`),
  serviceIntervals: () =>
    api<BackendServiceInterval[]>('/api/service-intervals'),
  recentMaintenance: (limit = 50) =>
    api<BackendMaintenanceRecord[]>(`/api/maintenance/recent?limit=${limit}`),
  recentFuelLogs: (limit = 100) =>
    api<BackendFuelLog[]>(`/api/fuel-logs/recent?limit=${limit}`),
  reminders: () =>
    api<MaintenanceReminder[]>('/api/maintenance/reminders'),
};
